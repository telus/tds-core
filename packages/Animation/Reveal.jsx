import React from 'react'
import PropTypes from 'prop-types'
import { Transition } from 'react-transition-group'

const defaultStyle = (timeout, delay) => {
  return {
    transition: `height ${timeout}ms ${delay ? `${delay}ms` : ''}`,
    overflow: 'hidden',
  }
}

const transitionStyles = height => ({
  entering: { height: `${height}px` },
  entered: { height: `${height}px`, overflow: 'visible' },
  exiting: { height: 0 },
  exited: { height: 0, visibility: 'hidden' },
})

const Reveal = ({ duration, timeout, height, delay, children, ...rest }) => {
  const transitionTimeout = duration || timeout
  return (
    <Transition {...rest} timeout={transitionTimeout + delay}>
      {status => (
        <div
          style={{
            ...defaultStyle(transitionTimeout, delay),
            ...transitionStyles(height)[status],
          }}
          aria-hidden={status === 'exited'}
          data-testid="childrenContainer"
        >
          {children()}
        </div>
      )}
    </Transition>
  )
}
Reveal.propTypes = {
  timeout: PropTypes.number.isRequired,
  delay: PropTypes.number,
  duration: PropTypes.number,
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  children: PropTypes.func.isRequired,
}

Reveal.defaultProps = {
  duration: undefined,
  delay: undefined,
}

export default Reveal
