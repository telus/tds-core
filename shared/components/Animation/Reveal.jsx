import React from 'react'
import PropTypes from 'prop-types'
import { Transition } from 'react-transition-group'

const defaultStyle = (duration, timeout) => ({
  transition: `height ${duration || timeout}ms`,
  height: 0,
  overflow: 'hidden',
})

const hiddenContent = {
  visibility: 'hidden',
}

const transitionStyles = height => {
  const styles = {
    height: `${height}px`,
  }

  return {
    entering: styles,
    entered: { height: 'auto' },
    exiting: styles,
    exited: hiddenContent,
  }
}

const Reveal = ({ height, children, ...rest }) => (
  <Transition {...rest}>
    {status => (
      <div
        style={{
          ...defaultStyle(rest.duration, rest.timeout),
          ...transitionStyles(height)[status],
        }}
        aria-hidden={status === 'exited'}
        data-testid="childrenContainer"
      >
        {children()}
      </div>
    )}
  </Transition>
)
Reveal.propTypes = {
  timeout: PropTypes.number.isRequired,
  duration: PropTypes.number,
  height: PropTypes.number.isRequired,
  children: PropTypes.func.isRequired,
}

Reveal.defaultProps = {
  duration: undefined,
}

export default Reveal
