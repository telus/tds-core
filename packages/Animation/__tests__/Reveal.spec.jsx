import React from 'react'
import { render, mount } from 'enzyme'

import Reveal from '../Reveal'

describe('Reveal', () => {
  const doMount = (overrides = {}) =>
    mount(
      <Reveal timeout={500} height={300} in={true} {...overrides}>
        {() => <div>Content to reveal</div>}
      </Reveal>
    )

  const doRender = (overrides = {}) =>
    render(
      <Reveal timeout={500} height={300} in={true} {...overrides}>
        {() => <div>Content to reveal</div>}
      </Reveal>
    )

  it('renders when the transition is triggered', () => {
    const reveal = doRender({ in: true })

    expect(reveal).toMatchSnapshot()
  })

  it('renders when the transition is not triggered', () => {
    const reveal = doRender({ in: false })

    expect(reveal).toMatchSnapshot()
  })

  it('marks the content as visible to screen readers when expanded', () => {
    const reveal = doMount({ in: true })

    expect(reveal.find('[data-testid="childrenContainer"]')).toHaveProp('aria-hidden', false)
  })

  it('marks the content as hidden to screen readers when contracted', () => {
    const reveal = doMount({ in: false })

    expect(reveal.find('[data-testid="childrenContainer"]')).toHaveProp('aria-hidden', true)
  })
})
