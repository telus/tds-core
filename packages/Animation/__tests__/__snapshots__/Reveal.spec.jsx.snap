// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reveal renders when the transition is not triggered 1`] = `
<div
  aria-hidden="true"
  data-testid="childrenContainer"
  style="transition:height 500ms;height:0;overflow:hidden;visibility:hidden"
>
  <div>
    Content to reveal
  </div>
</div>
`;

exports[`Reveal renders when the transition is triggered 1`] = `
<div
  aria-hidden="false"
  data-testid="childrenContainer"
  style="transition:height 500ms;height:auto;overflow:hidden"
>
  <div>
    Content to reveal
  </div>
</div>
`;
